### PR TITLE
Add metric_args

### PR DESCRIPTION
**Pull request by guarin**
## What has changed and why?

* Add `metric_args` to all train commands
* Unify metrics handling into `TaskMetric` classes
* Remove `metric_log_train`, `metric_log_classwise` fields from `model_args`
* Remove `metrics` fields from `model_args`
* Log metrics to console after each validation epoch

This PR is a collection of many smaller PRs to refactor metrics handling in LightlyTrain fine-tuning. This enables the following:
* Users can now configure metrics with `metric_args` which are validated with pydantic
* All models use a shared metrics API
* We can now properly track best and last validation metrics
* We can now properly show metrics summaries after each validation epoch
* Model training code is no longer mixed with ad-hoc metrics handling as all metrics code lives now in the `_metrics` module

The PR looks huge but it is essentially always the same changes:
* Define `MetricArgs`, `TaskMetricArgs`, and `TaskMetric` for each task (see new `_metrics` module)
* Use `TaskMetricArgs` and `TaskMetric` in all `train_model.py` files
* Update `train_task.py` and `train_task_helpers.py` to handle the new `TaskMetric` class

Overview of the metrics system (documented in `_metrics/task_metric.py`):
```
# ## Metrics Structure Overview
#
# The metrics system has three layers:
#
# 1. **MetricArgs** (Pydantic): Configuration for individual metrics
#    - Example: MulticlassAccuracyArgs(topk=[1, 5], average=["micro", "macro"])
#
# 2. **TaskMetricArgs** (Pydantic): Configuration for all metrics in a task
#    - Example: MulticlassClassificationTaskMetricArgs(
#         accuracy=..., f1=..., precision=...
#     )
#    - Groups related MetricArgs for a task. This is what the user configures
#
# 3. **TaskMetric** : Class that stores actual Metric instances
#    - Example: ClassificationTaskMetric(metric_args, class_names=[...], prefix="val_metric/")
#    - Stores actual torchmetrics instances (MetricCollection, MeanMetric, etc.)
#    - Provides .update(), .update_loss(), and .reset() for updating metrics during training/validation
#    - Provides .compute() for logging
#    - All metric logic for a task must be contained in this class. The training loop
#      only interacts with TaskMetric objects, not individual Metric instances.
#
```

## How has it been tested?

- [x] LTDETR DINOv3
- [x] LTDETR DINOv2
- [x] SemSeg DINOv3 EoMT
- [x] SemSeg DINOv2 EoMT
- [x] SemSeg Multihead
- [x] InstSeg DINOv3 EoMT
- [x] InstSeg DINOv2 EoMT
- [x] PanopticSeg DINOv3 EoMT
- [x] Classification
- [x] Classification Multihead

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [x] Yes
- [ ] Not needed (internal change)

## Did you update the documentation?

- [x] Yes
- [ ] Not needed (internal change without effects for user)

### CODE REVIEW

File: docs/source/settings/train_settings.md [hunk 1]
```diff
@@ -334,16 +335,14 @@ interruption.
...
-| [`save_best`](#save_best)                       | `bool`             | Track the best-performing checkpoint according to [`watch_metric`](#watch_metric).                    |
-| [`watch_metric`](#watch_metric)                 | `str`              | Validation metric name (for example `"val_metric/map"`) monitored when selecting the best checkpoint. |
-| [`mode`](#mode)                                 | `"min"`<br>`"max"` | Operation used when selecting the best checkpoint based on [`watch_metric`](#watch_metric).           |
+| Key                                             | Type   | Description                                                                                         |
+| ----------------------------------------------- | ------ | --------------------------------------------------------------------------------------------------- |
+| [`save_every_num_steps`](#save_every_num_steps) | `int`  | Training step interval for saving checkpoints.                                                      |
+| [`save_last`](#save_last)                       | `bool` | Persist `last.ckpt` after each save cycle. Disable only when storage is constrained.                |
```
Comment (liopeer) [inline docs/source/settings/train_settings.md:344]: What implications does this have on storage? How can I configure to save every epoch, i.e. `epoch1.ckpt`, `epoch2.ckpt`?

File: docs/source/settings/train_settings.md
Comment (liopeer) [inline docs/source/settings/train_settings.md]: Where to find them?

Comment (liopeer) [inline docs/source/settings/train_settings.md:180]: Having that inside `model_args` seems a bit arbitrary.

File: src/lightly_train/_metrics/loss_metrics.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        self.split = split
+        # For training we only track the losses of the most recent batches to keep an
+        # up-to-date estimate of the current loss. We do not only track the last loss
+        # value because of gradient accumulation which can cause the loss to fluctuate
+        # a lot from accumulation step to accumulation step.
+        metric_cls = (
+            partial(RunningMean, window=running_mean_window)
+            if split == "train"
+            else MeanMetric
+        )
```
Comment (guarin) [inline src/lightly_train/_metrics/loss_metrics.py]: @liopeer I went with the simple solution to use a constant window size of 20. This might slightly change the logged losses depending on the number of gradient accumulation steps but should overall be relatively stable. And it avoids having to pass the number of gradient accumulation steps to all models.

File: src/lightly_train/_metrics/task_metric.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        metrics:
+            Dictionary mapping metric names to computed float values.
+            For multihead metrics, contains per-head keys (e.g.,
+            "val_metric_head/miou_lr0_001") and best-head top-level keys
+            (e.g., "val_metric/miou").
+        watch_metric:
+            The key of the metric used for model selection
+            (e.g., "val_metric/miou"). Used by the training loop for
+            checkpointing. None if the watch metric is not found in the
+            computed metrics (e.g., a train metric watching a val key).
```
Comment (liopeer) [inline src/lightly_train/_metrics/task_metric.py:57]: Don't really understand this.

File: src/lightly_train/_commands/train_task.py [hunk 4]
```diff
@@ -1330,6 +1388,9 @@ def _train_task_from_config(config: TrainTaskConfig) -> None:
             "Contact us at https://www.lightly.ai/contact to discuss the best licensing option for your use case."
         )
 
+        # TODO(Guarin, 02/26): Add best metric to state?
```
Comment (liopeer) [inline src/lightly_train/_commands/train_task.py:1391]: That's actually a really good idea for continuing runs. Does not have to be done right now, but I would put it into an issue (if not already done).

File: src/lightly_train/_metrics/task_metric.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+
+def get_watch_metric_mode(
+    task_metric_args: TaskMetricArgs,
+    loss_names: list[str],
+    watch_metric_name: str,
+) -> Literal["min", "max"]:
+    """Determine whether watch_metric_name should be maximised or minimised.
+
+    Note 1: The matching is not exact key matching. Instead, we check if the
```
Comment (liopeer) [inline src/lightly_train/_metrics/task_metric.py:116]: Instead of this heuristic, I could in the future imagine to have containers for every metric that clearly define whether maximization/minimization is desired. So the collection of metrics would not be `dict[str, float]`, but `dict[str, Metric]` with:
```python
@dataclass
class Metric
    name: str
    mode: Literal["min", "max"]
```
Maybe I am also missing something why this would not be possible.

File: src/lightly_train/_commands/train_task_helpers.py [hunk 6]
```diff
@@ -926,123 +1079,67 @@ def add_timer_logs(
...
 
+    if last_metrics.watch_metric is None or last_metrics.watch_metric_value is None:
+        return best_metrics
 
-def average_accumulated_log_dict(
-    accumulated: dict[str, Any], num_accumulation_steps: int
-) -> None:
-    """Divide all plain Tensor values in accumulated by num_accumulation_steps in-place.
+    if best_metrics.metrics.watch_metric != last_metrics.watch_metric:
+        return best_metrics  # Shouldn't happen, but safe to assume we want best metrics
```
Comment (liopeer) [inline src/lightly_train/_commands/train_task_helpers.py:1118]: I'd consider raising an error here.

File: src/lightly_train/_metrics/task_metric.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+    def compute(self) -> MetricComputeResult:
+        """Compute all metrics and return values for logging."""
+        raise NotImplementedError
+
+    def reset(self) -> None:
+        """Reset all metrics."""
+        for module in self.modules():
+            if isinstance(module, Metric):
+                module.reset()
```
Comment (liopeer) [inline src/lightly_train/_metrics/task_metric.py:106]: From what I understand so far it would make more sense to me if those were inside `_task_models` and not inside `_metrics`.

File: src/lightly_train/_metrics/loss_metrics.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+from __future__ import annotations
+
+from collections.abc import Mapping, Sequence
+from functools import partial
+
+from torch import Tensor
+from torch.nn import Module, ModuleDict
+
+
+class LossMetrics(Module):
```
Comment (liopeer) [inline src/lightly_train/_metrics/loss_metrics.py:17]: Is there a reason why we don't inherit from `torchmetrics.Metric` here? EDIT: Is it because torchmetrics would reduce across ranks etc. which is undesired (we want to reduce gradients instead)?

File: src/lightly_train/_metrics/detection/task_metric.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+                    prefix=f"{split}_metric",
+                    class_names=class_names,
+                    iou_type="bbox",
+                    box_format=box_format,
+                )
+            )
+        self.metrics = MetricCollection(metrics)  # type: ignore
+        self.loss_metrics = LossMetrics(split=split, loss_names=loss_names)
+
+    def update(
```
Comment (liopeer) [inline src/lightly_train/_metrics/detection/task_metric.py:88]: I know it is not supposed to calculate the loss again, but the name `update` is super confusing and I would expect it to update everything. Can we include the `loss_dict` into the `update` method (instead of having a separate method? Or does torchmetrics' interface not allow for that?

File: src/lightly_train/_metrics/task_metric.py (hunk 7 above)
Comment (liopeer) [reply src/lightly_train/_metrics/task_metric.py:106]: Reasoning explained.

**Comment by guarin** [comment]
Will merge and create follow-up PRs for comments
